src/components/text-editor/prosemirror-adapter/plugins/trigger/factory.ts

```diff
@@ -0,0 +1,224 @@
+import { Plugin, PluginKey, EditorState, Transaction } from 'prosemirror-state';
+import { ReplaceStep, ReplaceAroundStep } from 'prosemirror-transform';
+import { EditorView } from 'prosemirror-view';
+import { TriggerCharacter, TriggerEventDetail } from './types';
+import { inserterFactory } from './inserter';
+
+const isTrigger = (
+    key: string,
+    validTriggers: TriggerCharacter[],
+): key is TriggerCharacter => {
+    return key.length === 1 && validTriggers.includes(key as TriggerCharacter);
+};
+
+const shouldTrigger = (state: EditorState): boolean => {
+    const { $from } = state.selection;
+
+    if ($from.pos === 1) {
+        return true;
+    }
+
+    // Getting the position immediately before the current selection
+    const prevPos = $from.pos - 1;
+
+    if (prevPos > 0) {
+        // allow trigger if the cursor is at the start of a new paragraph
+        if ($from.parentOffset === 0) {
+            return true;
+        }
+
+        const prevChar = state.doc.textBetween(prevPos, $from.pos);
+
+        return prevChar === ' ' || prevChar === '\n';
+    }
+
+    return false;
+};
+
+const stillHasTrigger = (
+    state: EditorState,
+    activeTrigger: string,
+    triggerPosition: number,
+    triggerLength: number,
+): boolean => {
+    const cursorPosition = state.selection.$from.pos;
+
+    if (
+        cursorPosition < triggerPosition ||
+        cursorPosition > triggerPosition + triggerLength + 1
+    ) {
+        return false;
+    }
+
+    return (
+        state.doc.textBetween(triggerPosition, triggerPosition + 1) ===
+        activeTrigger
+    );
+};
+
+const getTriggerEventDetail = (
+    view: EditorView,
+    trigger: TriggerCharacter,
+    value: string,
+): TriggerEventDetail => {
+    return {
+        trigger: trigger,
+        textEditor: inserterFactory(view),
+        value: value,
+    };
+};
+
+const sendTriggerEvent = (
+    type: 'triggerStart' | 'triggerStop' | 'triggerChange',
+    view: EditorView,
+    trigger: TriggerCharacter,
+    value: string,
+) => {
+    const event = new CustomEvent<TriggerEventDetail>(type, {
+        detail: getTriggerEventDetail(view, trigger, value),
+        bubbles: true,
+        composed: true,
+    });
+    view.dom.dispatchEvent(event);
+};
+
+const processTransactions = (
+    text: string,
+    transactions: Transaction[],
+    oldState: EditorState,
+): string => {
+    let textAdded = '';
+    let textRemoved = '';
+
+    transactions.forEach((transaction) => {
+        if (transaction.docChanged) {
+            transaction.steps.forEach((step) => {
+                if (
```

Comment:
Still some nested code, and also unnecessary inline code comments, but I'm ok merging it and get it in if it works as expected 😄